Handle when ALSA tweaks our sample spec so much that the frame size changes.

From: Pierre Ossman <ossman@fefdeb5f-60dc-0310-8127-8f9354f1896f>

(closes #57).

git-svn-id: svn://svn.0pointer.net/pulseaudio/trunk@1431 fefdeb5f-60dc-0310-8127-8f9354f1896f
---

 src/modules/module-alsa-sink.c |   11 +++++++++--
 1 files changed, 9 insertions(+), 2 deletions(-)

diff --git a/src/modules/module-alsa-sink.c b/src/modules/module-alsa-sink.c
index 6ff9a6e..df04281 100644
--- a/src/modules/module-alsa-sink.c
+++ b/src/modules/module-alsa-sink.c
@@ -155,8 +155,12 @@ static void do_write(struct userdata *u) {
             else
                 memchunk = &u->memchunk;
         }
-            
-        assert(memchunk->memblock && memchunk->memblock->data && memchunk->length && memchunk->memblock->length && (memchunk->length % u->frame_size) == 0);
+
+        assert(memchunk->memblock);
+        assert(memchunk->memblock->data);
+        assert(memchunk->length);
+        assert(memchunk->memblock->length);
+        assert((memchunk->length % u->frame_size) == 0);
 
         if ((frames = snd_pcm_writei(u->pcm_handle, (uint8_t*) memchunk->memblock->data + memchunk->index, memchunk->length / u->frame_size)) < 0) {
             if (frames == -EAGAIN)
@@ -412,6 +416,9 @@ int pa__init(pa_core *c, pa_module*m) {
         goto fail;
     }
 
+    /* ALSA might tweak the sample spec, so recalculate the frame size */
+    frame_size = pa_frame_size(&ss);
+
     if (ss.channels != map.channels)
         /* Seems ALSA didn't like the channel number, so let's fix the channel map */
         pa_channel_map_init_auto(&map, ss.channels, PA_CHANNEL_MAP_ALSA);
